File: backend/src/sandbox/overlay/writable_dirs.py
"""Writable overlay directory allocation.

Overlayfs needs a writable ``upperdir`` plus a sibling ``workdir`` for every
mounted overlay. Lower layers are leased from the layer stack; this module owns
only the upper/work side of the mount.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

OVERLAY_WRITABLE_ROOT = Path("/eos-mount-scratch/eos-sandbox-runtime")


class OverlayWritableRootUnavailable(RuntimeError):
    """Raised when the canonical upper/work root is not available."""


@dataclass(frozen=True)
class OverlayWritableDirs:
    """Per-overlay writable directories created beside each other."""

    run_dir: Path
    upperdir: Path
    workdir: Path


def overlay_writable_root() -> Path:
    """Return the canonical filesystem for overlay ``upperdir``/``workdir``.

    There is intentionally no fallback. Overlayfs requires upper/work dirs to
    live on a filesystem suitable for writable overlay state; Docker-backed
    sandboxes provide that filesystem at ``/eos-mount-scratch``.
    """
    root = OVERLAY_WRITABLE_ROOT
    if not root.is_dir():
        raise OverlayWritableRootUnavailable(
            f"overlay writable root is missing: {root}"
        )
    return root


def allocate_overlay_writable_dirs(run_dir: Path) -> OverlayWritableDirs:
    """Create and return the upper/work dirs for one overlay instance."""
    upperdir = run_dir / "upper"
    workdir = run_dir / "work"
    upperdir.mkdir(parents=True, exist_ok=True)
    workdir.mkdir(parents=True, exist_ok=True)
    return OverlayWritableDirs(run_dir=run_dir, upperdir=upperdir, workdir=workdir)


__all__ = [
    "OVERLAY_WRITABLE_ROOT",
    "OverlayWritableDirs",
    "OverlayWritableRootUnavailable",
    "allocate_overlay_writable_dirs",
    "overlay_writable_root",
]

File: sandbox/crates/eos-overlay/src/writable_dirs.rs
//! Writable overlay directory allocation.
//!
//! Overlayfs needs a writable `upperdir` plus a sibling `workdir` for every
//! mounted overlay. Lower layers are leased from the layer stack; this module
//! owns only the upper/work side of the mount. There is intentionally NO
//! fallback root — Docker-backed sandboxes provide the writable filesystem at
//! `/eos-mount-scratch`.

use std::path::{Path, PathBuf};

use crate::error::Result;

/// Canonical filesystem for overlay `upperdir`/`workdir`.
/// `// PORT backend/src/sandbox/overlay/writable_dirs.py:13 — OVERLAY_WRITABLE_ROOT`
pub const OVERLAY_WRITABLE_ROOT: &str = "/eos-mount-scratch/eos-sandbox-runtime";

/// Per-overlay writable directories created beside each other under one run dir.
/// `// PORT backend/src/sandbox/overlay/writable_dirs.py:20-26 — OverlayWritableDirs`
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct OverlayWritableDirs {
    /// The per-overlay run directory the upper/work dirs live under.
    pub run_dir: PathBuf,
    /// The overlay `upperdir` (`run_dir/upper`).
    pub upperdir: PathBuf,
    /// The overlay `workdir` (`run_dir/work`).
    pub workdir: PathBuf,
}

/// Return the canonical writable root, creating it if its parent exists.
///
/// Mirrors Python `overlay_writable_root()`: create `OVERLAY_WRITABLE_ROOT`
/// only when its parent is already a directory, then require the result to be a
/// directory or raise [`OverlayError::WritableRootUnavailable`]. No fallback.
/// `// PORT backend/src/sandbox/overlay/writable_dirs.py:29-43 — overlay_writable_root`
pub fn overlay_writable_root() -> Result<PathBuf> {
    // PORT backend/src/sandbox/overlay/writable_dirs.py:36-42 — mkdir-if-parent then is_dir gate
    todo!()
}

/// Create and return the `upper`/`work` dirs for one overlay instance.
/// `// PORT backend/src/sandbox/overlay/writable_dirs.py:46-52 — allocate_overlay_writable_dirs`
pub fn allocate_overlay_writable_dirs(run_dir: &Path) -> Result<OverlayWritableDirs> {
    // PORT backend/src/sandbox/overlay/writable_dirs.py:48-52 — mkdir upper/work (parents, exist_ok)
    let _ = run_dir;
    todo!()
}
